File: ch4/UVa10004.ml
open Scanf
open Printf

let solve n m =
  let adj = Array.make n [] in
  for i = 1 to m do
    let u, v = scanf "%d %d\n" (fun x y -> (x, y)) in
    adj.(u) <- v :: adj.(u);
    adj.(v) <- u :: adj.(v)
  done;
  let color = Array.make n (-1) in
  let q = Queue.create () in
  Queue.add 0 q;
  let is_bipartite = ref true in
  while not (Queue.is_empty q) && !is_bipartite do
    let u = Queue.pop q in
    adj.(u) |> List.iter (fun v ->
      if color.(v) = -1 then begin
        color.(v) <- 1 - color.(u);
        Queue.add v q
      end else if color.(u) = color.(v) then
        is_bipartite := false
    )
  done;
  printf "%sBICOLORABLE.\n" (if !is_bipartite then "" else "NOT ")

let () =
  let rec loop () =
    try
      let n, m = scanf "%d\n%d\n" (fun x y -> (x, y)) in
      solve n m;
      loop ()
    with _ -> () in
  loop ()

File: ch4/UVa10004.cpp
// Bicoloring

#include <bits/stdc++.h>
using namespace std;

#define INF 1e9

typedef vector<int> vi;

int main() {
  int n;
  while (scanf("%d", &n), n) {
    vector<vi> AL(n, vi());
    int l; scanf("%d", &l);
    while (l--) {
      int a, b; scanf("%d %d", &a, &b);
      AL[a].push_back(b);
      AL[b].push_back(a); // bidirectional
    }
    int s = 0;
    queue<int> q; q.push(s);
    vi color(n, INF); color[s] = 0;
    bool isBipartite = true;   // add one more boolean flag, initially true
    while (!q.empty() && isBipartite) { // as with the original BFS routine
      int u = q.front(); q.pop();
      for (auto &v : AL[u]) {
        if (color[v] == INF) {       // but, instead of recording distance,
          color[v] = 1-color[u];        // we just record two colors {0, 1}
          q.push(v);
        }
        else if (color[v] == color[u]) {       // u & v have the same color
          isBipartite = false;               // we have a coloring conflict
          break;
        }
      }
    }
    printf("%sBICOLORABLE.\n", (isBipartite ? "" : "NOT "));
  }
  return 0;
}
